src/acc/pack_kernel.cpp

```diff
@@ -0,0 +1,329 @@
+/*
+ Crown Copyright 2012 AWE.
+
+ This file is part of CloverLeaf.
+
+ CloverLeaf is free software: you can redistribute it and/or modify it under
+ the terms of the GNU General Public License as published by the
+ Free Software Foundation, either version 3 of the License, or (at your option)
+ any later version.
+
+ CloverLeaf is distributed in the hope that it will be useful, but
+ WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ details.
+
+ You should have received a copy of the GNU General Public License along with
+ CloverLeaf. If not, see http://www.gnu.org/licenses/.
+ */
+
+//  @brief Fortran mpi buffer packing kernel
+//  @author Wayne Gaudin
+//  @details Packs/unpacks mpi send and receive buffers
+
+#include "pack_kernel.h"
+#include "context.h"
+
+void clover_pack_message_left(global_variables &globals, int x_min, int x_max, int y_min, int y_max, clover::Buffer2D<double> &field_buffer,
+                              clover::Buffer1D<double> &left_snd_buffer, int cell_data, int vertex_data, int x_face_data, int y_face_data,
+                              int depth, int field_type, int buffer_offset) {
+
+  // Pack
+
+  int x_inc = 0, y_inc = 0;
+
+  // These array modifications still need to be added on, plus the donor data location changes as in update_halo
+  if (field_type == cell_data) {
+    x_inc = 0;
+    y_inc = 0;
+  }
+  if (field_type == vertex_data) {
+    x_inc = 1;
+    y_inc = 1;
+  }
+  if (field_type == x_face_data) {
+    x_inc = 1;
+    y_inc = 0;
+  }
+  if (field_type == y_face_data) {
+    x_inc = 0;
+    y_inc = 1;
+  }
+
+  // DO k=y_min-depth,y_max+y_inc+depth
+
+  double *left_snd = left_snd_buffer.data;
+  double *field = field_buffer.data;
+  const size_t field_sizex = field_buffer.nX();
+/* TODO NVHPC Compiler is also saying that it can't parallelize this loop
```

Comment:
This one is slightly more concerning, does `collapse(2)` help? 
One of the OpenMP target implementations we've tested (it may have been vanilla Clang) had some performance issues with `collapse(2)`  in packing/unpacking so we didn't write it that way.

Reply:
I am not exactly sure what the impact of `collapse(2)` will have on performance. I would probably need to do profiling to fully investigate. It does pass validation with `clover_bm16_short.in` if I do add a `collapse(2)`. 

We could add the `collapse(2)` just for OpenACC since NVHPC seems to be the only compiler with a mature implementation. 

As for the OpenMP version maybe there are performance issues with vanilla Clang due to `simd` not being fully implemented in `teams distribute parallel for simd`? I think we should add the `collapse(2)` to the OpenMP version if I am assuming the NVHPC compiler does not have performance issues. Hopefully vanilla clang's OpenMP implementation progresses where this won't become an issue in the future?

Reply:
Yes, please profile and keep `collapse(2)` if it helps. Open a separate PR if you think OpenMP will also benefit from this and the compiler doesn't fail or given wrong results.